cms/analysis/models.py

```diff
@@ -0,0 +1,226 @@
+from functools import cached_property
+from typing import TYPE_CHECKING, Any, ClassVar, Optional, cast
+
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.http import Http404
+from django.shortcuts import redirect
+from django.utils.translation import gettext_lazy as _
+from wagtail.admin.panels import FieldPanel, FieldRowPanel, HelpPanel, MultiFieldPanel, TitleFieldPanel
+from wagtail.contrib.routable_page.models import RoutablePageMixin, path
+from wagtail.fields import RichTextField
+from wagtail.models import Page
+from wagtail.search import index
+
+from cms.analysis.blocks import AnalysisStoryBlock
+from cms.core.blocks import HeadlineFiguresBlock
+from cms.core.fields import StreamField
+from cms.core.models import BasePage
+
+if TYPE_CHECKING:
+    from django.http import HttpRequest
+    from django.http.response import HttpResponseRedirect
+    from django.template.response import TemplateResponse
+    from wagtail.admin.panels import Panel
+
+
+class AnalysisSeries(RoutablePageMixin, Page):
+    """The analysis series model."""
+
+    parent_page_types: ClassVar[list[str]] = ["topics.TopicPage"]
+    subpage_types: ClassVar[list[str]] = ["AnalysisPage"]
+    preview_modes: ClassVar[list[str]] = []  # Disabling the preview mode due to it being a container page.
+    page_description = _("A container for Analysis series")
+
+    content_panels: ClassVar[list["Panel"]] = [
+        *Page.content_panels,
+        HelpPanel(
+            content="This is a container for Analysis series. It provides the <code>/latest</code>,"
+            "<code>/previous-release</code> evergreen paths, as well as the actual analysis pages. "
+            "Add a new Analysis page under this container."
+        ),
+    ]
+
+    def get_latest(self) -> Optional["AnalysisPage"]:
```

Comment:
We use forward references (`"AnalysisPage"`) and Python doesn't support that syntax (https://bugs.python.org/issue45857)

```
$ mypy cms
...
    def get_latest(self) -> "AnalysisPage" | None:
                            ~~~~~~~~~~~~~~~^~~~~~
TypeError: unsupported operand type(s) for |: 'str' and 'NoneType'
```

One option is to do the whole thing as `def get_latest(self) -> "AnalysisPage | None"` which I am in two minds about.

Reply:
`Class | None` is the preferred shorhand. `Optional[Foo]` is equivalent to `Union[Foo, None]` and I prefer the Optional format in this case, to be honest

Reply:
> We use forward references (`"AnalysisPage"`) and Python doesn't support that syntax ([bugs.python.org/issue45857](https://bugs.python.org/issue45857))
> 
> ```
> $ mypy cms
> ...
>     def get_latest(self) -> "AnalysisPage" | None:
>                             ~~~~~~~~~~~~~~~^~~~~~
> TypeError: unsupported operand type(s) for |: 'str' and 'NoneType'
> ```
> 
> One option is to do the whole thing as `def get_latest(self) -> "AnalysisPage | None"` which I am in two minds about.

Interesting, I didn't know `|` didn't support forward references. 👍 

Re: Union, | vs Optional.

We can update our style guide to reflect that, I do like optional, but I think we should use None over Optional if there is more than 1 type.

So to clarify this is my suggestion, let me know your thoughts:

```py
# single type that is optional
def get_latest(self) -> Optional[AnalysisPage]:
	...

# multiple types which is optional

# not preferred - confusing even though it is functionally the same
def get_latest(self) -> Optional[AnalysisPage] | OtherPage
	...

# preferred
def get_latest(self) -> AnalysisPage | OtherPage | None
	...

```

I guess the only time we make an exception for Optional with multiple types is if they use forward references. Or we use Union["AnalysisPage", "OtherPage"] for those.

Thoughts?. Or do people prefer stringifying the whole thing `"AnalysisPage | OtherPage | None"`?

Reply:
I am all in for `def get_latest(self) -> AnalysisPage | OtherPage | None` with 2 exceptions:
1. `Optional["ForwardReference"]
2. `Union["ForwardReference", NonForwardReference]` / `Union["ForwardReference", NonForwardReference, None]` (`None` is a non-forward ref, but this is to make it more explicit for the `"AnalysisPage" | OtherPage | None` use case

Reply:
So, to confirm, we allow Union/Optional if any one type has a forward reference.

For Optional, I am happy to use that for single types or we stick with `|` unless forward references are involved.

```
# this
def get_latest(self) -> Optional[AnalysisPage]:
	...

# instead of
def get_latest(self) -> Analysis | None
	...
```

Thoughts?

Reply:
aligned with you on this one

Reply:
I'll update our style guide accordingly.